# Pre-processing and Modeling

In [1]:
# import libraries
import pandas as pd
import numpy as np

In [2]:
# read the cleaned movies data set
movies = pd.read_csv('/Users/Atabay/Desktop/Movie_Recommendation_System_data/Data/m_data_cleaned.csv')

In [3]:
movies.head()

budget                                             genres      id  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   19995   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...     285   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...  206647   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   49026   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                                production_countries release_date  runtime  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10    162.0   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19    169.0   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26    148.0   
3  [{"iso_3166_1": "US", "name": "United States o...   2012-07-16    165.0   
4  [{"iso_3166_1": "US", "name": "United States o...   2012-03-07    132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466  
3                     The Dark Knight Rises           7.6        9106  
4                               John Carter           6.1        2124

In [4]:
movies.shape

(4802, 17)

In [5]:
credits = pd.read_csv('/Users/Atabay/Desktop/Movie_Recommendation_System_data/Data/c_data_cleaned.csv')

In [6]:
credits.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

## Demographic Filtering

In the demographic filtering, I will use vote average as the only decisive factor to rank the movies. However, some movies have very low vote count which would make the ranking unfair. To solve the issue I will only include the movies with considerable vote avarage to my ranking.

I will use 99 percentile as my cut off for vote count because we have 44035 movies in our data and most of have very few votes which may cause ranking to be biased. 

In [26]:
m_90 = movies['vote_count'].quantile(0.90)
m_90

1839.2000000000044

In [27]:
r_90 = movies.copy().loc[movies['vote_count'] >= m_90]
r_90.shape

(481, 17)

We have 481 movies that have voting cote more than 1839. I will use only those movies in my ranking.

In [28]:
r_90 = r_90.sort_values('vote_average', ascending=False)

In [29]:
# Displaying top 20 movies
r_90[['title', 'vote_count', 'vote_average']].head(20)

title  vote_count  vote_average
1881         The Shawshank Redemption        8205           8.5
3337                    The Godfather        5893           8.4
2294                    Spirited Away        3840           8.3
662                        Fight Club        9413           8.3
1818                 Schindler's List        4329           8.3
3865                         Whiplash        4254           8.3
3232                     Pulp Fiction        8428           8.3
2731           The Godfather: Part II        3338           8.3
4601                     12 Angry Men        2078           8.2
690                    The Green Mile        4048           8.2
809                      Forrest Gump        7927           8.2
1847                       GoodFellas        3128           8.2
1987             Howl's Moving Castle        1991           8.2
1990          The Empire Strikes Back        5879           8.2
65                    The Dark Knight       12002           8.2
2247                Princess Mononoke        1983           8.2
2170                           Psycho        2320           8.2
3719  One Flew Over the Cuckoo's Nest        2919           8.2
3057               American History X        3016           8.2
2453               Dead Poets Society        2705           8.1

This ranking can be used for simple movie recommendation to users. It is not influenced by any user preference or choice therefore It will be basic and same for everyone.

## Content Based Filtering

### a. Plot based recommender 

I will build a content based filtering using similarity scores of words in overview column

In [11]:
movies.overview.head()

0    In the 22nd century, a paraplegic Marine is di...
1    Captain Barbossa, long believed to be dead, ha...
2    A cryptic message from Bond’s past sends him o...
3    Following the death of District Attorney Harve...
4    John Carter is a war-weary, former military ca...
Name: overview, dtype: object

In [12]:
# Importing the neccessary library
from sklearn.feature_extraction.text import TfidfVectorizer

In [13]:
# Defining a tdif vectorizer and removing all the english words stop words
tfidf = TfidfVectorizer(stop_words='english')

In [15]:
#Constructing the required tfidf matrix by fitting and transforming the data
tfidf_matrix = tfidf.fit_transform(movies['overview'])

In [16]:
tfidf_matrix.shape

(4802, 20980)

It seems that there are over 20980 words to describe 4802 movies

In [17]:
# Importing linear_kernel
from sklearn.metrics.pairwise import linear_kernel

# Compute the cosine similarity matrix
cosine_sim = linear_kernel(tfidf_matrix, tfidf_matrix)

We need to define a function taking movie title as an input and recommending 5 similar movies as an output

In [18]:
#Creating a reverse map of indexes and movie titles
indexes = pd.Series(movies.index, index=movies['title']).drop_duplicates()

In [19]:
indexes

title
Avatar                                         0
Pirates of the Caribbean: At World's End       1
Spectre                                        2
The Dark Knight Rises                          3
John Carter                                    4
                                            ... 
El Mariachi                                 4797
Newlyweds                                   4798
Signed, Sealed, Delivered                   4799
Shanghai Calling                            4800
My Date with Drew                           4801
Length: 4802, dtype: int64

#### Defining the recommendation function

In [20]:
# Function that takes in movie title as input and outputs the most similar movies
def get_recommendations(title, cosine_sim=cosine_sim):
    # Index of the movie that matches the title
    index = indexes[title]

    # Pairwsie similarity scores of all movies with that movie
    sim_scores = list(enumerate(cosine_sim[index]))

    # Sort the movies based on the similarity scores
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    # Get the scores of the 5 most similar movies
    sim_scores = sim_scores[1:6]

    # Get the movie indixes
    movie_indexes = [i[0] for i in sim_scores]

    # Return the top 10 most similar movies
    return movies['title'].iloc[movie_indexes]

In [62]:
get_recommendations('Avatar')

3604               Apollo 18
2130            The American
634               The Matrix
1341    The Inhabited Island
529         Tears of the Sun
Name: title, dtype: object

In [22]:
get_recommendations('Spirited Away')

1825            Jimmy Neutron: Boy Genius
4599                      51 Birch Street
131                               G-Force
2394                               Wolves
117     Charlie and the Chocolate Factory
Name: title, dtype: object

The recommendation system does a good job returning movies with similar plot descriptions. However, we can increase the quality of the recommendation system by adding more metadata.

### b. Adding More Data to Recommendation Engine

I will be using cast and crew columns to extract the 3 top actors, and director info. Also, I will be using genre info to improve my recommendation engine. 

In [34]:
credits.columns = ['id','title_credits','cast','crew']

In [35]:
credits.head()

id                             title_credits  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [36]:
# Merging movies and credits data
df = movies.merge(credits,on='id')

In [37]:
df.head()

budget                                             genres      id  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   19995   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...     285   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...  206647   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   49026   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                                production_countries release_date  runtime  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10    162.0   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19    169.0   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26    148.0   
3  [{"iso_3166_1": "US", "name": "United States o...   2012-07-16    165.0   
4  [{"iso_3166_1": "US", "name": "United States o...   2012-03-07    132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                               John Carter           6.1        2124   

                              title_credits  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", 

Our data is in the string form we need to convert it into usable structure.

In [38]:
# Reorganizing the string features 
from ast import literal_eval

features = ['cast', 'crew', 'genres']
for feature in features:
    df[feature] = df[feature].apply(literal_eval)

In [43]:
# From the crew column getting the director's name. If director is not found, return NaN
def get_director(x):
    for i in x:
        if i['job'] == 'Director':
            return i['name']
    return np.nan


In [44]:
# Returning the top 3 element from a list
def get_list(x):
    if isinstance(x, list):
        names = [i['name'] for i in x]
        #Check if more than 3 elements exist. If yes, return only first three. If no, return entire list.
        if len(names) > 3:
            names = names[:3]
        return names

    #Return empty list in case of missing data
    return []

In [45]:
# Get director from crew column and get top 3 elements of cast and genre columns
df['director'] = df['crew'].apply(get_director)
 
features = ['cast', 'genres']
for feature in features:
    df[feature] = df[feature].apply(get_list)

In [48]:
# Get an overview of the new features
df[['title', 'cast', 'director', 'genres']].head()

title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                               cast           director  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]      James Cameron   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]     Gore Verbinski   
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]         Sam Mendes   
3      [Christian Bale, Michael Caine, Gary Oldman]  Christopher Nolan   
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]     Andrew Stanton   

                                 genres  
0          [Action, Adventure, Fantasy]  
1          [Adventure, Fantasy, Action]  
2            [Action, Adventure, Crime]  
3                [Action, Crime, Drama]  
4  [Action, Adventure, Science Fiction]

We need to strip and lowercase names to avoid confusion between the values having the same first names.

In [50]:
# Striping blank spaces and converting to lower case 
def clean_data(x):
    if isinstance(x, list):
        return [str.lower(i.replace(" ", "")) for i in x]
    else:
        #Checking if director exists, if not return empty string
        if isinstance(x, str):
            return str.lower(x.replace(" ", ""))
        else:
            return ''

In [51]:
# Apply clean_data function to your features.
features = ['cast', 'director', 'genres']

for feature in features:
    df[feature] = df[feature].apply(clean_data)

In [55]:
def movie_soup(x):
    return ' '.join(x['cast']) + ' ' + x['director'] + ' ' + ' '.join(x['genres'])
df['soup'] = df.apply(movie_soup, axis=1)

Applying Count Vectorizer instead of tfidf because word frequency is important for our engine. We don't need to penalize most frequent words.

In [57]:
# Import CountVectorizer and create the matrix
from sklearn.feature_extraction.text import CountVectorizer

count = CountVectorizer(stop_words='english')
c_matrix = count.fit_transform(df['soup'])

In [58]:
# Computing the Cosine Similarity matrix based on the c_matrix
from sklearn.metrics.pairwise import cosine_similarity

cosine_sim2 = cosine_similarity(c_matrix, c_matrix)

In [59]:
# Resetting index of our main DataFrame and creating reverse mapping
df = df.reset_index()
indixes = pd.Series(df.index, index=df['title'])

Comparing the two recommendation engine I created

In [64]:
get_recommendations('Avatar', cosine_sim)

3604               Apollo 18
2130            The American
634               The Matrix
1341    The Inhabited Island
529         Tears of the Sun
Name: title, dtype: object

In [63]:
get_recommendations('Avatar', cosine_sim2)

206                         Clash of the Titans
1      Pirates of the Caribbean: At World's End
5                                  Spider-Man 3
9            Batman v Superman: Dawn of Justice
10                             Superman Returns
Name: title, dtype: object

All five movie recommendations are different for the engines. Adding more metadata to our engine led to better results.